src/box/alter.cc

```diff
@@ -3871,6 +3871,75 @@ on_replace_dd_priv(struct trigger * /* trigger */, void *event)
 
 /* {{{ cluster configuration */
 
+/**
+ * This trigger implements the "last write wins" strategy for
+ * "bootstrap_leader_uuid" tuple of space _schema. Comparison is performed by
+ * a timestamp and replica_id of the node which authored the change.
+ */
+static int
+before_replace_dd_schema(struct trigger * /* trigger */, void *event)
+{
+	struct txn *txn = (struct txn *)event;
+	struct txn_stmt *stmt = txn_current_stmt(txn);
+	struct tuple *old_tuple = stmt->old_tuple;
+	struct tuple *new_tuple = stmt->new_tuple;
+	const char *key = tuple_field_cstr(new_tuple ? new_tuple : old_tuple,
+					   BOX_SCHEMA_FIELD_KEY);
+	if (key == NULL)
+		return -1;
+	if (strcmp(key, "bootstrap_leader_uuid") == 0) {
+		uint64_t old_ts = 0;
+		uint32_t old_id = 0;
+		uint64_t new_ts = UINT64_MAX;
+		uint32_t new_id = UINT32_MAX;
+		/*
+		 * Assume anything can be stored in old_tuple, so do not require
+		 * it to have a timestamp or replica_id. In contrary to that,
+		 * always require new tuple to have a valid timestamp and
+		 * replica_id.
+		 */
+		if (old_tuple != NULL) {
+			const char *field = tuple_field(old_tuple, 2);
```

Comment:
I tried to do this, but stumbled upon a problem:
`_schema`, unlike other system spaces, holds arbitrary key - value pairs, and occasionally multiple values.

It's more or less clear, that the second field should be called `BOX_SCHEMA_FIELD_VALUE` (and @Gerold103 has already done so in his instance names patch).
As for fields 3 and 4 - they can't share a common name.
For example, for "version" tuple (['version', 2, 11, 1]), fields 3 and 4 are something like `BOX_SCHEMA_FIELD_VERSION_MINOR`, `BOX_SCHEMA_FIELD_VERSION_PATCH`.

And for my new tuple ['bootstrap_leader_uuid', 'c0488250-f2d8-4c5e-9706-7cb70c4e434e', 1681746623593755, 1] fields 3 and 4 would be `BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP` and `BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID`.

Long story short, the enum names should be specific to the particular _schema tuple, and this way they become too long.

On the other hand, I could make fields 3 and 4 names `BOX_SCHEMA_FIELD_3` and `BOX_SCHEMA_FIELD_4`.
But such names are no better than using plain numbers, it seems.

I'm open to your suggestions on how to shorten the enum names. It would be better to access the fields by a proper name rather than some magic number, indeed.

I think we can change the naming schema for "_schema" tuples. There will be generic `BOX_SCHEMA_FIELD_KEY` and `BOX_SCHEMA_FIELD_VALUE`. And also `BOX_VERSION_FIELD_MAJOR/MINOR/PATCH`, `BOX_BOOTSTRAP_LEADER_UUID/TIMESTAMP/ID`.

Reply:
enums can hold several names for the same value, maybe we can have something like:
```c
/** _schema fields. */
enum {
        BOX_SCHEMA_FIELD_KEY = 0,
        BOX_SCHEMA_FIELD_VALUE = 1,
...
        BOX_SCHEMA_FIELD_VERSION_MINOR = 3,
        BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP = 3,
        BOX_SCHEMA_FIELD_VERSION_PATCH = 4,
        BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID = 4,
...
};
```

Reply:
Yes, but what about the length?
```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index 0de262b4c..5ce546b1d 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -3900,16 +3900,22 @@ before_replace_dd_schema(struct trigger * /* trigger */, void *event)
                 * replica_id.
                 */
                if (old_tuple != NULL) {
-                       const char *field = tuple_field(old_tuple, 2);
+                       const char *field = tuple_field(old_tuple,
+                               BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP);
                        if (field != NULL && mp_typeof(*field) == MP_UINT)
                                old_ts = mp_decode_uint(&field);
-                       field = tuple_field(old_tuple, 3);
+                       field = tuple_field(old_tuple,
+                               BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID);
                        if (field != NULL && mp_typeof(*field) == MP_UINT)
                                old_id = mp_decode_uint(&field);
                }
                if (new_tuple != NULL &&
-                   (tuple_field_u64(new_tuple, 2, &new_ts) != 0 ||
-                    tuple_field_u32(new_tuple, 3, &new_id) != 0)) {
+                   (tuple_field_u64(new_tuple,
+                                    BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP,
+                                    &new_ts) != 0 ||
+                    tuple_field_u32(new_tuple,
+                                    BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID,
+                                    &new_id) != 0)) {
                        return -1;
                }
                if (new_ts < old_ts || (new_ts == old_ts && new_id < old_id)) {
```
I can't fit such long names anywhere. And tuple_field access which took a single line now takes 2 or 3.

Omitting the _SCHEMA part (like BOX_FIELD_BOOTSTRAP_LEADER_TIMESTAMP) is still too long.

Reply:
Ok, I checked other existing long enum names starting with `BOX_SCHEMA_*`, and in comparison to them our new names are not drastically longer:
```
BOX_FUNC_FIELD_IS_DETERMINISTIC                                                  
BOX_FK_CONSTRAINT_FIELD_PARENT_COLS                                              
BOX_FUNC_INDEX_FIELD_SPACE_ID                                                    
BOX_SESSION_SETTINGS_FIELD_VALUE                                                 
BOX_SPACE_SEQUENCE_FIELD_IS_GENERATED                                            

BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_UUID                                           
BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP                                      
BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID
```

I tried to shorten the names, but failed to find any good alternative.
I think it's better to stick with existing naming schema then. Will do.

P. S. Maybe @Gerold103 has some thoughts on how to shorten the `BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_*`?

Reply:
I wouldn't care too much and just used `BOX_SCHEMA_FIELD_VALUE + 1`, `BOX_SCHEMA_FIELD_VALUE + 2`, etc. But if you anyway want new names, then either use the long name, or maybe try `BOX_SCHEMA_FIELD_BOOT_LEADER_`, without `STRAP` suffix. I can't think of any shorter alternative.

Reply:
@sergepetrenko I'm ok with `BOX_SCHEMA_FIELD_VALUE + <number>`

Reply:
Done:
```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index 7c743652e..c129b1467 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -3893,6 +3893,8 @@ before_replace_dd_schema(struct trigger * /* trigger */, void *event)
                uint32_t old_id = 0;
                uint64_t new_ts = UINT64_MAX;
                uint32_t new_id = UINT32_MAX;
+               int ts_field = BOX_SCHEMA_FIELD_VALUE + 1;
+               int id_field = BOX_SCHEMA_FIELD_VALUE + 2;
                /*
                 * Assume anything can be stored in old_tuple, so do not require
                 * it to have a timestamp or replica_id. In contrary to that,
@@ -3900,16 +3902,16 @@ before_replace_dd_schema(struct trigger * /* trigger */, void *event)
                 * replica_id.
                 */
                if (old_tuple != NULL) {
-                       const char *field = tuple_field(old_tuple, 2);
+                       const char *field = tuple_field(old_tuple, ts_field);
                        if (field != NULL && mp_typeof(*field) == MP_UINT)
                                old_ts = mp_decode_uint(&field);
-                       field = tuple_field(old_tuple, 3);
+                       field = tuple_field(old_tuple, id_field);
                        if (field != NULL && mp_typeof(*field) == MP_UINT)
                                old_id = mp_decode_uint(&field);
                }
                if (new_tuple != NULL &&
-                   (tuple_field_u64(new_tuple, 2, &new_ts) != 0 ||
-                    tuple_field_u32(new_tuple, 3, &new_id) != 0)) {
+                   (tuple_field_u64(new_tuple, ts_field, &new_ts) != 0 ||
+                    tuple_field_u32(new_tuple, id_field, &new_id) != 0)) {
                        return -1;
                }
                if (new_ts < old_ts || (new_ts == old_ts && new_id < old_id)) {
```